rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+- there are some OS syscall to do in order to enable expanded stack
+  space.
+
+
+There are two options here:
+- expose two APIs, `brgemm::set_hw_context()` and
+  `brgemm::release_hw_context()`. Those can be scheduled/hoisted as
+  the user see fit, but the setup function will have to be called
+  before the brgemm::execute call for each given brgemm object. The
+  release call can happen anytime the user see fit.
+- expose only `brgemm::release_hw_context()`, and hide context
+  setting. Those setting calls would have to be hidden under existing
+  functionalities (e.g. brgemm::execute). This slightly simplifies the
+  programming model, though it removes some flexibility on the user
+  side to hoist those setting calls (e.g. to do them only once per
+  thread for whole process lifetime).
+- don't expose those functionalities. This is very risky since this
+  can impact the performance of the overall application by lowering
+  the frequency for pieces of code that don't rely on those matrix
+  accelerators.
+
+The recommendation is to go with the first option, with explicit set
+and release functions, that the user can hoist as they see fit.
+
+With respect to OS syscall, we recommend to make it transparent to
+user, by making those upon first created brgemm object that would need
+to use the matrix acceleration engine. This will ensure this call
+happens before any call to hardware instruction requiring it.
+
+## Handling of attributes
+
+First, here is the list of attributes we plan on exposing:
+- fp_math mode, in order for user to control input conversions semantics;
+- scales/zero-point in order to support upconversion before computation.
+- post-ops, in order to allow fusion of elementwise and binary
+  operations.
+
+Internally, other attributes are available to improve performance, but
+given how they rely on implicit knowledge of brgemm implementation
+internals, it is preferable to not expose those as of now.
+
+We also have internal attributes to control prefetch hints, prefetch
+distances and non-temporal stores. However those are not yet fully
+supported by current brgemm kernel implementation so there is no plan
+to expose those for now.
+
+API wise, block level attributes can be exposed in two ways:
+- relying on existing oneDNN primitive attribute. This allows to
+  simplify documentation and usability by aligning semantics between
+  primitive and block level APIs. However, there is a set of
+  attributes that are specific to primitive API, and we might expose a
+  set of attributes specific to the block-level API (e.g. prefetching
+  strategy).
+- relying on dedicated attributes for block-level API. This allows a
+  clear separation of attributes between block level and primitive
+  level APIs, and allows to expose only supported attributes for block
+  level API.
+
+
+To make a first version of the API publicly available faster,
+we will start by relying on existing dnnl::primitive_attr objects.
```

Comment:
I guess starting with a separate attribute abstraction would leave more flexibility to revise in the future?